client/service/src/builder.rs

```diff
@@ -679,21 +679,27 @@ impl<TBl, TRtApi, TCl, TFchr, TSc, TImpQu, TFprb, TFpp, TExPool, TRpc, Backend>
 			finality_proof_request_builder: self.finality_proof_request_builder,
 			finality_proof_provider: self.finality_proof_provider,
 			transaction_pool: Arc::new(transaction_pool),
-			rpc_extensions: self.rpc_extensions,
+			rpc_extensions_builder: self.rpc_extensions_builder,
 			remote_backend: self.remote_backend,
 			block_announce_validator_builder: self.block_announce_validator_builder,
 			marker: self.marker,
 		})
 	}
 
 	/// Defines the RPC extensions to use.
-	pub fn with_rpc_extensions<URpc>(
+	pub fn with_rpc_extensions<URpcBuilder, URpc>(
 		self,
-		rpc_ext_builder: impl FnOnce(&Self) -> Result<URpc, Error>,
-	) -> Result<ServiceBuilder<TBl, TRtApi, TCl, TFchr, TSc, TImpQu, TFprb, TFpp,
-		TExPool, URpc, Backend>, Error>
-	where TSc: Clone, TFchr: Clone {
-		let rpc_extensions = rpc_ext_builder(&self)?;
+		rpc_ext_builder: impl FnOnce(&Self) -> Result<URpcBuilder, Error>,
+	) -> Result<
+		ServiceBuilder<TBl, TRtApi, TCl, TFchr, TSc, TImpQu, TFprb, TFpp, TExPool, URpc, Backend>,
+		Error,
+	>
+	where
+		TSc: Clone,
+		TFchr: Clone,
+		URpcBuilder: Fn(sc_rpc::DenyUnsafe) -> URpc + Send + 'static,
```

Comment:
We could consider traitifying `Fn(sc_rpc::DenyUnsafe) -> URpc + Send + 'static` and implement that trait for both closures and concrete `URpc` type, so that we could maintain backward compatibility. Not sure if it's worth it though.

Reply:
This might be a good idea actually, for example in the light client currently we don't care about `DenyUnsafe` for any rpc extensions. I guess right now anyone who isn't using BABE's RPC wouldn't care that much about it either.

Reply:
@tomusdrw how does one do that?  Every time I have tried to do that it fails with some obscure lifetime error.

Reply:
You mean this: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=59c9d81d289603e857e5a9dbb8970486 ?